spec/java/util/HashSet.main.lsl

```diff
@@ -0,0 +1,633 @@
+        var resultArray: array<Object> = action ARRAY_NEW("java.lang.Object", this.length);
+        var i: int = 0;
+
+        if (aLen < size)
+        {
+            action LOOP_FOR(
+                i, 0, size, +1,
+                toArray_loop(i, visitedKeys, resultArray)
+            );
+        }
+        else
+        {
+            resultArray = a;
+            action LOOP_FOR(
+                i, 0, size, +1,
+                toArray_loop(i, visitedKeys, resultArray)
+            );
+
+            if (aLen > this.length)
+                result[this.length] = null;
+        }
+
+        _checkForComodification(expectedModCount);
+        result = resultArray;
+    }
+
+
+    fun *.containsAll (@target self: HashSet, c: Collection): boolean
+    {
+        val otherSize: int = action CALL_METHOD(c, "size", []);
+        val iter: Iterator = action CALL_METHOD(c, "iterator", []);
+        var isContainsAll: boolean = true;
+
+        action LOOP_WHILE(
+            action CALL_METHOD(iter, "hasNext", []),
+            _containsAllElements_loop(iter, isContainsAll)
+        );
+
+        result = isContainsAll;
+    }
+
+
+    @Phantom proc _containsAllElements_loop(iter: Iterator, isContainsAll: boolean): void
+    {
+        val key: Object = action CALL_METHOD(iter, "next", []);
+        val isKeyExist: boolean = action MAP_HAS_KEY(this.storage, key);
+
+        if (!isKeyExist)
+        {
+            isContainsAll = false;
+            action LOOP_BREAK();
+        }
+    }
+
+
+    fun *.addAll (@target self: HashSet, c: Collection): boolean
+    {
+        result = _addAllElements(c);
+    }
+
+
+    fun *.retainAll (@target self: HashSet, c: Collection): boolean
+    {
+        if (c == null)
+            _throwNPE();
+
+        val lengthBeforeAdd: int = this.length;
+        val iter: Iterator = action CALL_METHOD(c, "iterator", []);
+
+        action LOOP_WHILE(
+            action CALL_METHOD(iter, "hasNext", []),
+            _retainAllElements_loop(iter)
+        );
+
+
+        if (lengthBeforeAdd != this.length)
+        {
+            this.modCount += 1;
+            result = true;
+        }
+        else
+        {
+            result = false;
+        }
+    }
+
+
+    @Phantom proc _retainAllElements_loop(iter: Iterator): void
+    {
+        val key: Object = action CALL_METHOD(iter, "next", []);
+        val hasKey: boolean = action MAP_HAS_KEY(this.storage, key);
+
+        if (!hasKey)
+        {
+            action MAP_REMOVE(this.storage, key);
+            this.length -= 1;
+        }
+    }
+
+
+    fun *.removeIf (@target self: HashSet, filter: Predicate): boolean
+    {
+        if (filter == null)
+            _throwNPE();
+
+        val lengthBeforeAdd: int = this.length;
+        var i: int = 0;
+        val visitedKeys: map<Object, Object> = action MAP_NEW();
+
+
+        action LOOP_WHILE(
+            i < lengthBeforeAdd,
+            _removeIf_loop(i, visitedKeys, filter)
```

Comment:
Should there be `modCount` checks present anywhere?

Reply:
Fixed